lib/internal/modules/esm/hooks.js

```diff
@@ -115,27 +115,46 @@ class Hooks {
       },
     ],
   };
+}
+
+class Hooks {
+  #chains;
 
   // Cache URLs we've already validated to avoid repeated validation
-  #validatedUrls = new SafeSet();
+  #validatedUrls;
+
+  constructor(chains = getDefaultChains(), validatedUrls = new SafeSet()) {
+    this.#chains = chains;
+    this.#validatedUrls = validatedUrls;
+  }
 
   /**
    * Import and register custom/user-defined module loader hook(s).
    * @param {string} urlOrSpecifier
    * @param {string} parentURL
    */
   async register(urlOrSpecifier, parentURL) {
-    const moduleLoader = require('internal/process/esm_loader').esmLoader;
-
-    const keyedExports = await moduleLoader.import(
+    const esmLoader = require('internal/process/esm_loader').esmLoader;
+    const keyedExports = await esmLoader.import(
       urlOrSpecifier,
       parentURL,
       kEmptyObject,
     );
-
     this.addCustomLoader(urlOrSpecifier, keyedExports);
   }
 
+  allowImportMetaResolve() {
```

Comment:
I don't particularly like these three functions… but in particular `getChains` and `getValidatedUrls` seem like they should be done another way.